ENH: add a lung mask segmentation node selector in Parenchyma Analysis

rbumm opened the pull request:
- enable the segmentation node selector if input volume is selected
- convert the segmentation to a labelmap using CIP color conventions
- polulate the labelmap selector

only works if segmentations for "right lung", "left lung" and "other" are detected in the node

rbumm committed:
ENH add a lung mask segmentation node selector

- moved segmentation to labelmap conversion to CIP_ParenchymaAnalysisLogic
- new parameter "segmentationNode" CIP_ParenchymaAnalysisLogic
- if present, convert to temporary labelmap
- cleanup from widget because temporary labelmapNode is needed for statistics

lassoan commented on Scripted/CIP_ParenchymaAnalysis/CIP_ParenchymaAnalysis.py:
```diff
@@ -133,6 +133,22 @@ def setup(self):
+        #
+        # the segmentation node selector
+        #
+        self.segmentationSelector = slicer.qMRMLNodeComboBox()
```
Use the `self.labelSelector` for selecting both labelmap volume nodes and segmentation nodes.

```
self.labelSelector.nodeTypes = ["vtkMRMLLabelMapVolumeNode", "vtkMRMLSegmentationNode"]
```

Having two node selectors would complicate the GUI and could lead to ambiguities (what happens if you select both a segmentation and a labelmap volume node?).

lassoan commented on Scripted/CIP_ParenchymaAnalysis/CIP_ParenchymaAnalysis.py:
```diff
@@ -329,6 +348,43 @@ def onCTSelect(self, node):
+                                       "Parenchyma Analysis", "Input segmentation not valid.")
+                return False
+            else: 
+                segmentToLabelValueMapping = slicer.util.getFirstNodeByClassByName("vtkMRMLColorTableNode", "CIP colors")
```
I've noticed that CIP uses multiple different color tables. Rename `CIP colors` to `chest_region_colors_basic` (to indicate that it is derived from [chest_region_colors.ctbl](https://github.com/acil-bwh/SlicerCIP/blob/develop/Scripted/CIP_/CIP/ui/Resources/chest_region_colors.ctbl)). Change the name in LungCTAnalyzer, too, if needed (but hopefully we'll not need to define color tables there if we do the conversion to labelmap here).

If the color table is not needed for displaying the results then you can remove both the temporary labelmap and color node after the execution is completed, to not litter the scene with temporary nodes.

lassoan reviewed (commented):
Thank you, this will be great! I've added a few comments inline.

Note for the future: There are at least 3 different color tables here and label value definitions are scattered throughout SlicerCIP and ChestImagingPlatform at dozens of places. It would be hard to harmonize all these and get rid of the hardcoded label values, so the approach that is implemented here (convert segmentation to labelmaps using label values that the specific module expects) could work well. We would just need to move this conversion to some utility classes that can be accessed from multiple modules so that we don't need to repeat code. But this can be done when we want to add segmentation node support to more modules.

rbumm commented:
> Thank you, this will be great! I've added a few comments inline.
> 
> Note for the future: There are at least 3 different color tables here and label value definitions are scattered throughout SlicerCIP and ChestImagingPlatform at dozens of places. It would be hard to harmonize all these and get rid of the hardcoded label values, so the approach that is implemented here (convert segmentation to labelmaps using label values that the specific module expects) could work well. We would just need to move this conversion to some utility classes that can be accessed from multiple modules so that we don't need to repeat code. But this can be done when we want to add segmentation node support to more modules.

I made the change to a uniform lung mask selector which accepts labelmap and segmentation node. 

I renamed the color table in all places. 

Was not able to delete the temp labelmap and the color table at the end of onApply ()  because if the user presses "Apply" multiple times you get error messages because the temp labelmap is missing for the new analysis. Pls advise. These are only two files, maybe one could just leave them in the scene?

rbumm commented:
Unfortunately, this produced another pull request. What should I have done instead and what should I do now resp tomorrow (good night)?

lassoan commented:
The proper solution is to keep pushing your changes to the same branch (rbumm/develop) and they show up automatically here. Even if you have modified change history (e.g., amended an existing commit, squashed commits that were already pushed) you can still push and overwrite all previous content on a branch, you just need to force-push (in TortoiseGit push dialog: "Force: May discard known changes").

I would recommend to try to force-push your latest version of the to this branch and if you verified that you see the correct code here then close the other pull request.

lassoan commented on Scripted/CIP_ParenchymaAnalysis/CIP_ParenchymaAnalysis.py:
```diff
@@ -133,22 +133,24 @@ def setup(self):
         # the label map volume selector
         #
         self.labelSelector = slicer.qMRMLNodeComboBox()
         # self.labelSelector.nodeTypes = ( ("vtkMRMLScalarVolumeNode"), "" )
```
You can delete these two commented-out lines.

lassoan commented on Scripted/CIP_ParenchymaAnalysis/CIP_ParenchymaAnalysis.py:
```diff
@@ -330,22 +332,61 @@ def onCTSelect(self, node):
+                            self.segmentToLabelValueMapping.SetColor(58, "trachea",    0.49, 0.49, 0.79, 1.0)
+                            self.segmentToLabelValueMapping.NamesInitialisedOn()
+
+                        self.labelNode = slicer.util.getFirstNodeByClassByName("vtkMRMLLabelMapVolumeNode",self.segmentationNode.GetName() + "_LabelMap")
```
Always create a new labelmap volume node and a new color table node and delete them after the processing is completed. It does not take much time to create these nodes from scratch and it is simpler and safer then reusing nodes.

lassoan reviewed (commented):
It mostly looks good, just added two comments inline.

lassoan commented:
I see that you have pushed some changes. Let me know if the code is ready for review.

rbumm commented:
Sry - should be ready for review, Andras -  the code now removes the temporary files from segmentation conversion (only after conversion is really done) and these files are now always created newly. Tested a few CTs and it works well.

lassoan commented on Scripted/CIP_ParenchymaAnalysis/CIP_ParenchymaAnalysis.py:
```diff
@@ -330,22 +330,58 @@ def onCTSelect(self, node):
+                          segmentIds.InsertNextValue(segmentId)
+                        slicer.modules.segmentations.logic().ExportSegmentsToLabelmapNode(self.segmentationNode, segmentIds, self.labelNode, \
+                            self.CTNode, slicer.vtkSegmentation.EXTENT_REFERENCE_GEOMETRY, self.segmentToLabelValueMapping)
+                        self.tempLabelMap = True
```
There is no need for this flag. If the temporary color table or the temporary labelmap node is not set to None then we remove them (and set them to None). We also need to initialize them to None.

rbumm replied:
OK, will be changed tomorrow.

lassoan commented on Scripted/CIP_ParenchymaAnalysis/CIP_ParenchymaAnalysis.py:
```diff
@@ -330,22 +330,58 @@ def onCTSelect(self, node):
-            for color in ['Red', 'Yellow', 'Green']:
-                slicer.app.layoutManager().sliceWidget(color).sliceLogic().GetSliceCompositeNode().SetLabelVolumeID('None')
-
+        if not node: 
```
The segmentation to labelmap conversion has to happen in the logic, because if you want to use this module from another module then you cannot use anything that is in the module widget class. You can report error message from the logic by raising an exception and catching it in the widget class.

rbumm replied:
OK, logical, will be changed tomorrow.

lassoan reviewed (commented):
Thanks a lot for the updates, it is getting there. I added just a few comments. I hope I'm not asking too much.

rbumm commented:
The segmentation to labelmap conversion now happens in the logic. I needed to define a new parameter "segmentationNode" in the logic. If this is present, the conversion takes place and overrides a "labelNode" parameter. Cleanup had to be done by call from from the widget because the widget needs the labelmap for the late statistics. Quite complicated ...

rjosest commented:
@rbumm  Thank you so much for making these changes. They look good from my end although I am not very familiar with the new segmentation infrastructure.
@lassoan I totally agree about the need for unified util for colormap handling. We devise our internal color nodes because the CIP segmentation tools follow a label map convention that is specific to CIP. The least significant 8 bits encode the concept of "Region" and the 8 most significant bits encodes the concept of "Type".

If nobody objects, I can proceed with the merge of the pull request

rbumm commented:
Maybe @lassoan should give his final ok. 

@rjosest I enjoyed working on this and use the Airway Segmenter and Parenchyma analysis quite often.

rjosest commented:
@rbumm Glad to hear that CIP is useful for your research. Feel to share any improvements

lassoan commented:
I'll have a look at the code later this afternoon, you could hold off the merge until then.

rbumm commented:
@lassoan just a reminder here thanks

rbumm commented:
@lassoan this probably got forgotten - could you check the code if you find the time? Thank you.

lassoan commented:
Thanks for the reminder. I've been working on a regression in the Lung CT segmentation module caused by the update of Markups module last week. I've managed to fix that problem yesterday and will get back to review and test this code today.

rbumm commented:
Closed, because now included in #34 .

rbumm commented:
> I've been working on a regression in the Lung CT segmentation module caused by the update of Markups module

@lassoan there is still a problem in Lung CT Segmenter concerning markup placement. green slice constantly switches back to red slice after adding the 7th marker (Preview Slicer 4.13.0-2021-09-28)
